prow/cmd/deck/static/prow/prow.ts

```diff
@@ -697,3 +848,121 @@ function createIcon(iconString: string, tooltip: string = ""): HTMLAnchorElement
+    public state: JobState;
+    public row: number;
+
+    constructor(start: number, duration: number, state: JobState, row: number) {
+        this.start = start;
+        this.duration = duration;
+        this.state = state;
+        this.row = row;
+    }
+}
+
+export class JobHistogram {
+    public start: number;
+    public end: number;
+    private data: JobSample[];
+
+    constructor() {
+        this.data = [];
+        this.start = Number.MAX_SAFE_INTEGER;
+        this.end = 0;
+    }
+    // add adds a sample to the histogram, filtering states that didn't result in success or clear
+    // failure, and updating the range of the histogram data.
+    public add(sample: JobSample) {
+        if (!(sample.state === "success" || sample.state === "failure" || sample.state === "error")) {
+            return;
+        }
+        if (sample.start < this.start) {
+            this.start = sample.start;
+        }
+        if (sample.start > this.end) {
+            this.end = sample.start;
+        }
+        this.data.push(sample);
+    }
+    // buckets assigns all samples between start and end into cols buckets, sorted by
+    // start timestamp, while the buckets themselves are sorted by duration.
+    public buckets(start: number, end: number, cols: number): JobBuckets {
+        this.data.sort((s1, s2) => s1.start - s2.start);
+        const stride = (end - start) / cols;
+        const buckets: JobSample[][] = [];
+
+        let max = 0;
+        let next = start;
+        this.data.forEach((sample) => {
+            if (sample.start < start || sample.start > end) {
+                return;
+            }
+            if (sample.duration > max) {
+                max = sample.duration;
+            }
+            if (sample.start < next || sample.start === end) {
+                buckets[buckets.length - 1].push(sample);
+                return;
+            }
+            if (buckets.length > 0) {
+                const bucket = buckets[buckets.length - 1];
+                bucket.sort((s1, s2) => s1.duration - s2.duration);
+            }
+            next = next + stride;
+            while (next < sample.start) {
+                buckets.push([]);
+                next = next + stride;
+            }
+            buckets.push([sample]);
+        });
+        if (buckets.length > 0) {
+            const lastBucket = buckets[buckets.length - 1];
+            lastBucket.sort((s1, s2) => s1.duration - s2.duration);
+        }
+        while (buckets.length < cols) {
+            buckets.push([]);
+        }
+        return new JobBuckets(buckets, start, end, max);
+    }
+    // length returns the number of samples in the histogram.
+    public length(): number {
+        return this.data.length;
+    }
+}
+
+export class JobBuckets {
+    public data: JobSample[][];
+    public start: number;
+    public end: number;
+    public max: number;
+
+    constructor(data: JobSample[][], start: number, end: number, max: number) {
```

Comment:
As above, you could also write this constructor like so to repeat yourself less:

```ts
export class JobBuckets {
    constructor(public data: JobSample[][],
                public start: number,
                public end: number,
                public max: number) { }

    // ...
}
```